### ISSUE TITLE
issue with training AOD-Net

### ISSUE THREAD

**Issue by YogeshShitole** [sachinpuranik99/AOD_Net#1]
Hi @sachinpuranik99 
Thanks for redirecting me from here [https://github.com/Boyiliee/AOD-Net/issues/2](url)  to your AOD-Net train script
I tried training it with NYU2 database with 27, 256 images as mentioned in AOD paper
I downloaded training(hazy images) set and original image(clear image) set from there portal
for each clear image in original set  training set has multiple synthetically generated hazy images

My problem is during training my loss is in order of 10^7 and final accuracy after 27256 iteration is ~64% (I already trained it once for all 27256 images), I was facing similar issue with my caffe train script with test_template.prototxt file available at  Boyliiee/AOD-Net repo    [ https://github.com/Boyiliee/AOD-Net/issues/5](url)
can you tell me what may be the resoan

# snapshot from my training

 5937/27256 [=====>........................] - ETA: 1:50:55 - loss: 6875409.1251 - acc: 0.6050X: ['AOD_train/training/NYU2_1160_1_1.jpg']
   Label--------------------------------------------------------------------- ['AOD_train/original/NYU2_1160.jpg']

 5938/27256 [=====>........................] - ETA: 1:50:55 - loss: 6874253.4533 - acc: 0.6050X: ['AOD_train/training/NYU2_645_7_2.jpg']
   Label---------------------------------------------------------------------- ['AOD_train/original/NYU2_645.jpg']

please suggest if I am doing anything wrong since from your training log loss is very less and accuracy after training is more than 90%

# snippet from your aod_net.ipynb
  % delta_t_median)
4881/6500 [=====================>........] - ETA: 2:29 - loss: 0.0255 - acc: 0.9319- ETA: 2:30 - loss: 0.0255 - acc: 0.931/home/sachin/anaconda2/lib/python2.7/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.115872). Check your callbacks.
  % delta_t_median)
6500/6500 [==============================] - 592s 91ms/step - loss: 0.0252 - acc: 0.9314 - ETA:  - ETA: 1s - loss: 0.025

**Comment by sachinpuranik99** [comment]
Hi Yogesh, 

Thanks for trying out my code. I am not sure what exactly is missing in your code, as I have tried it on multiple datasets and it used to converge to 90+% accuracy in a moderate number of epochs.  

One thing I am noticing from your log is that 

> 5937/27256 [=====>........................] - ETA: 1:50:55 - loss: 6875409.1251 


The training is abruptly stopping at around 5937/5938 images, maybe that could be the issue, as the model is being trained on different images on individual epochs and not getting proper time to converge. 

I suggest you to do two things:
First, try maybe a subset of images first (say 1000) and see if it's converging.
Second, see if the train input and its corresponding labels are properly matched during the training phase.

Let me know the results.

HTH.

**Comment by YogeshShitole** [comment]
Hi @sachinpuranik99,

Thanks for quick response  :+1: 

# I trained network with all 27256 image below is  final output



27255/27256 [============================>.] - ETA: 0s - loss: 1502418.3378 - acc: 0.6684X: ['AOD_train/training/NYU2_705_1_2.jpg']
   Label------------------------------------------------------------ ['AOD_train/original/NYU2_705.jpg']

27256/27256 [=======================] - 8482s 311ms/step - loss: 1502363.2445 - acc: 0.6684

Process finished with exit code 0


# Then as you suggested I trained it with small subset of 1000 images from same dataset.. loss seems to reduce but accuracy is not good below is snippet from training on subset data 

 1/1000 [..............................] - ETA: 6:16 - loss: 675708800.0000 - acc: 0.9064X: ['AOD_train/subset/NYU2_481_5_2.jpg']
   Label------------------------------------------------------------------- ['AOD_train/original/NYU2_481.jpg']

   2/1000 [..............................] - ETA: 5:43 - loss: 957947072.0000 - acc: 0.8566X: ['AOD_train/subset/NYU2_597_1_1.jpg']
   Label------------------------------------------------------------------- ['AOD_train/original/NYU2_597.jpg']

                                                             -
                                                             -
                                                             -
998/1000 [============================>.] - ETA: 0s - loss: 41109794.3659 - acc: 0.7054X: ['AOD_train/subset/NYU2_33_2_1.jpg']
   Label-------------------------------------------------------------------- ['AOD_train/original/NYU2_33.jpg']

 999/1000 [============================>.] - ETA: 0s - loss: 41068660.5003 - acc: 0.7056X: **['AOD_train/subset/NYU2_14_2_3.jpg']**
   Label------------------------------------------------------------------- **['AOD_train/original/NYU2_14.jpg']**

1000/1000 [==============================] - 315s 315ms/step - loss: 41027609.1131 - acc: 0.7058

Process finished with exit code 0

# regarding weather train image and label is properly matched I am printing out label image path and train image path during training as you can see in snippet it is highlighted for every hazy image NYU2_11_(*)_(*).jpg there is clear NYU2_11.jpg image available in label directory

so I am very sure train and label images are matched but **not sure how loss is getting calculated** in my case  since your loss is coming in decimal but mine is order of 10^6 (It is strange to me is there any preprocessing step is required for input images and labels before feeding it for training like normalization) For current training I have not made any changes in your network definition... just changed train_data_dir and label_data_dir path.

I feel other reason may be it is due to nature of dataset it is not converging which will be surprising if true since I am using training data from original paper...
but to check If possible can you share your dataset with which you trained your network I will check if after training it is converging to 90% if not then may be some problem with my system and setup  ;(

**Comment by sachinpuranik99** [comment]
Hello Yogesh,

I had used the dataset from the below website to train my network. 
[RESIDE](https://sites.google.com/view/reside-dehaze-datasets/reside-standard?authuser=0). Follow the ITS and OTS links to download the data. 

On a different note, can you please point me to the NYU2 dataset? As far as I know, NYU2 dataset contains depth related information along with GT images. For our task, we require hazy images which are synthetically generated using the depth information.

**Comment by YogeshShitole** [comment]
Hi Sachin 
AOD paper used indoor NYU2 dtaset for training they took RGB images from there and synthetically generated hazy image by uniformly choosing Atmospheric light bw [0.6, 1.0] and other some parameters (refer to 'datasets and implementation' section in paper page 4)...
below I am sharing link for original ground truth images and hazy training images
for more details on dataset you can follow Boyliee's project page mentioned in paper
[https://sites.google.com/site/boyilics/website-builder/project-page](url)

indoor NYU2 AOD
original image
[https://drive.google.com/file/d/1Sz5ZFFZXo3sY85R3v7yJa6W6riDGur46/view](url)

training Images (27256 as mentioned in paper)
[https://drive.google.com/file/d/17ZWJOpH1AsYQhoqpWR6PK61HrUhArdAK/view](url)

**Comment by sachinpuranik99** [comment]
Ok, I downloaded the above dataset and tried to train the network. I was able to achieve the following results.

> Epoch 1/2
> 27256/27256 [==============================] - 692s 25ms/step - loss: 1160.9226 - acc: 0.8270
> Epoch 2/2
> 25053/27256 [==========================>...] - ETA: 1:23 - loss: 1073.3400 - acc: 0.8285 ETA: 1:23 - loss: 1073.3407 - acc: 0.82
> 

I am not sure what exactly is missing in your modified code. I am attaching the NYU2 specific modified code [here.](https://drive.google.com/file/d/1PeWNi_J4_NG4_csTEhye8Un1tlO4eGt6/view?usp=sharing) Please try to run my code first and then compare to see whats missing. Let me know if you have any other queries.

**Comment by YogeshShitole** [comment]
@sachinpuranik99 Hi Sachin,
sorry for late response, I tried your NYU2 specific modified code below is my result

```
Epoch 1/2
27256/27256 [==============================] - 9066s 333ms/step - loss: 1349.2202 - acc: 0.7807

Epoch 2/2
27255/27256 [============================>.] - ETA: 0s - loss: 1138.8420 - acc: 0.7868
27256/27256 [==============================] - 9273s 340ms/step - loss: 1138.8203 - acc: 0.7868
```

in my case accuracy is coming around 0.78 and loss: 1138
but another major issue is when I am using final  model
from above training instead of producing dehazed image it is producing completely black image...

**Comment by sachinpuranik99** [comment]
Which method did you use for saving the image? Also, which utility you are using for viewing the image?
Please elaborate. 
Regarding the training accuracy, did you change any part of the code or the dataset? If not, there should not be any reason for the 4% accuracy drop.

**Comment by YogeshShitole** [comment]
Hi Sachin,
I haven't modified your code I ran it as it is......
I just added one line after `    model.fit_generator(generator=train_generator(train_X, label_path, batch_size), steps_per_epoch=steps_per_epoch, epochs=epochs,verbose=1 )`  to save model like this `model.save(filepath='pretrained_model/NYU2/trainedNYU_model.h5')`

for testing using saved model
`   model = load_model('pretrained_model/NYU2/trainedNYU_model.h5',
                       custom_objects={
                           'relu': relu,
                       }
                       )`
 I used SOTS testing section of your code and tested it on SOTS combined(indoor+outdoor)
(clear , hazy) pair of image so saving method for dehazed image is also same as mentioned in SOTS testing section  `def save_image(Y_predicted_array, filenames):` and for showing image I used `def show_image(file_data):`

although when I run program  before training starts I get below message but it did not affect training program execution

`/home/ubuntu/anaconda2/lib/python2.7/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
2018-06-15 09:55:44.101552: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-15 09:55:44.101580: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-15 09:55:44.101585: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2018-06-15 09:55:44.101589: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-15 09:55:44.101593: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
`

**Comment by BrainPotter** [comment]
Hi YogeshShitole,
I noticed that you wrote training programs of Boyiliee's AOD-Net, and I want to train this program on my own dataset. Can you please share your all training programs to me?Thank you very much!
My email address is junnian_L@163.com

**Comment by sachinpuranik99** [comment]
Hello @BrainPotter, you can find the aod_net.ipnb notebook in this repo. The notebook contains the necessary code to train the network. You just need to change the datapath if you want to train on your own data. Let me know if you find any issues.
